eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/EnactSpec.hs

```diff
@@ -79,145 +100,343 @@ spec =
-        getsNES newEpochStateGovStateL
-
-      impAnn "Votes are recorded in the proposals" $ do
-        let proposalsWithVotes =
-              proposalsAddVote
-                (CommitteeVoter committeeMember)
-                VoteYes
-                govActionId
-                ( proposalsAddVote
-                    (DRepVoter dRep)
-                    VoteYes
-                    govActionId
-                    proposalsBeforeVotes
+      getTreasury `shouldReturn` initialTreasury
+      sumRewardAccounts withdrawals `shouldReturn` zero
+
+    it "Wthdrawals exceeding treasury submitted in several proposals within the same epoch" $ do
+      (drepC, committeeC, _) <- electBasicCommittee
+      modifyPParams $ ppGovActionLifetimeL .~ EpochInterval 5
+      initialTreasury <- getTreasury
+      -- generate withdrawals which individually do not exceed the treasury, but their sum does
+      numWithdrawals <- choose (1, 10)
+      withdrawals <- genWithdrawalsExceeding initialTreasury numWithdrawals 50
+
+      impAnn "submit in the same proposal, with no effect on the treasury" $ do
+        gaId <- submitTreasuryWithdrawals withdrawals
+        submitYesVote_ (DRepVoter drepC) gaId
+        submitYesVote_ (CommitteeVoter committeeC) gaId
+        passNEpochs 2
+        getTreasury `shouldReturn` initialTreasury
+
+      impAnn "submit in individual proposals in the same epoch" $ do
+        traverse_
+          ( \w -> do
+              gaId <- submitTreasuryWithdrawals @era [w]
+              submitYesVote_ (DRepVoter drepC) gaId
+              submitYesVote_ (CommitteeVoter committeeC) gaId
+          )
+          withdrawals
+        passNEpochs 2
+
+        let expectedTreasury =
+              foldl'
+                ( \acc (_, x) ->
+                    if acc <-> x >= zero
+                      then acc <-> x
+                      else acc
                 )
-        proposalsAfterVotes `shouldBe` proposalsWithVotes
+                initialTreasury
+                withdrawals
 
-      impAnn "EnactState has not changed" $
-        enactStateAfterVotes `shouldBe` enactStateBeforeVotes
+        getTreasury `shouldReturn` expectedTreasury
+        -- check that the sum of the rewards matches what was spent from the treasury
+        sumRewardAccounts withdrawals `shouldReturn` (initialTreasury <-> expectedTreasury)
+  where
+    getTreasury = getsNES (nesEsL . esAccountStateL . asTreasuryL)
+    sumRewardAccounts withdrawals = mconcat <$> traverse (getRewardAccountAmount . fst) withdrawals
+    genWithdrawalsExceeding (Coin val) n maxExcess = do
+      accounts <- replicateM n $ registerRewardAccount @era
+      pcts <- replicateM n (choose (1, 100) :: ImpTestM era Integer)
+      let tot = sum pcts
+      excess <- choose (1, maxExcess) :: ImpTestM era Integer
+      let amounts = fmap (\x -> Coin $ ceiling ((x + excess) % tot * fromIntegral val)) pcts
+      pure $ zip accounts amounts
```

Comment:
`maxExcess` is overused in here and its meaning is totally unclear. Why would every random `pct` (which I assume stands for percent, but isn't really a percent) include that excess? So, IMHO we don't need to exceed every account by a random 50 to 100 points, or whatever they may mean. It is enough that we have at least one account that exceeds by a tiny bit, it will be enough. Excess can be decided at random too:
```suggestion
    genWithdrawalsExceeding (Coin val) n = do
      pcts <- replicateM (n - 1) $ choose (1, 100)
      let tot = sum pcts
      Positive excess <- arbitrary
      let amounts = excess : map (\x -> Coin $ ceiling ((x * val) % tot)) pcts
      forM amounts $ \ amount -> (,amount) <$> registerRewardAccount
```

Reply:
I made the "excess" configurable for the test case above
``` generate withdrawals which individually do not exceed the treasury, but their sum does ```
but I agree i didn't make it clear what it means or how to use it.  I thought it's handy to control the generation in this way, but maybe it's a little excessive.

Reply:
I understand the goal, I am saying that it was unnecessary to achieve the goal.
Instead of incrementing every amount by some arbitrary percentage, it is better to add one amount that puts the whole sum over the top.

Reply:
Ok